src/crypto/PublicKey.ts

```diff
@@ -211,6 +217,123 @@ export class SignedPublicKey
+        obj.staticSignature
+      )
+    } else {
+      throw new Error(
+        'Unimplemented signature type for AccountLinkedPublicKeyV1'
+      )
+    }
+  }
+
+  // Return bytes of the encoded unsigned key.
+  private bytesToSign(): Uint8Array {
+    return this.keyBytes
+  }
+
+  public getLinkedAddress(role: AccountLinkedRole): string {
+    if (this.staticSignature) {
+      // Decode from UTF-8
+      const signatureText = new TextDecoder().decode(this.staticSignature.text)
+      if (
+        signatureText !==
+        WalletSigner.accountLinkRequestText(this.bytesToSign(), role)
+      ) {
+        throw new Error('Signature text does not match expected text')
+      }
+      const digest = hexToBytes(utils.hashMessage(signatureText))
+      const signature = this.staticSignature.walletEcdsaCompact
+      const publicKey = ecdsaSignerKey(digest, signature)
+      if (!publicKey) {
+        throw new Error('Static signature is not valid for given role')
+      }
+      return publicKey.getEthereumAddress()
+    } else {
+      throw new Error('SIWE signature not implemented')
+    }
+  }
+
+  // Verify that signature was created from the digest using matching private key.
+  verify(signature: Signature, digest: Uint8Array): boolean {
+    if (!signature.ecdsaCompact) {
+      return false
+    }
+    return secp.verify(
+      signature.ecdsaCompact.bytes,
+      digest,
+      this.secp256k1Uncompressed.bytes
+    )
+  }
+
+  // Verify that the provided public key was signed by matching private key.
+  async verifyKey(pub: PublicKey | SignedPublicKey): Promise<boolean> {
+    if (!pub.signature) {
+      return false
+    }
+    const digest = await sha256(pub.bytesToSign())
+    return this.verify(pub.signature, digest)
+  }
+}
+
+export class AccountLinkedPublicKey
+  extends AccountLinkedPublicKeyV1
+  implements publicKey.AccountLinkedPublicKey
+{
+  v1: publicKey.AccountLinkedPublicKey_V1
+
+  constructor(obj: publicKey.AccountLinkedPublicKey) {
+    if (obj.v1) {
+      super(obj.v1)
+      this.v1 = obj.v1
+    } else {
+      throw new Error('Unsupported AccountLinkedPublicKey version')
+    }
+  }
+
+  public static fromLegacyKey(
+    legacyKey: SignedPublicKey,
+    role: AccountLinkedRole
+  ): AccountLinkedPublicKey {
+    if (!legacyKey.keyBytes || !legacyKey.signature) {
+      throw new Error('key is not signed')
+    }
+    const textBytes = toUtf8Bytes(
+      WalletSigner.accountLinkRequestText(legacyKey.keyBytes, role)
+    )
+    const staticSignature = AccountLinkedStaticSignature.create(
+      textBytes,
+      legacyKey.signature
+    )
+    const key = new AccountLinkedPublicKey({
+      v1: new AccountLinkedPublicKeyV1({
+        keyBytes: legacyKey.keyBytes,
+        staticSignature,
+        siweSignature: undefined,
+      }),
+    })
+    if (!key.getLinkedAddress(role)) {
+      throw new Error('Legacy key has invalid signature for given role')
```

Comment:
Same comment as above, I could be wrong but I think it'll return some random address rather than throw or return null